src/connector/src/sink/kinesis.rs

```diff
@@ -191,39 +190,176 @@ impl KinesisSinkWriter {
+                                    warn!(err_string = ?result_entry.error_message, ?throttle_delay, "throttle");
+                                    sleep(throttle_delay).await;
+                                } else  {
+                                    // no progress due to some internal error
+                                    assert_eq!(first_failed_idx, 0);
+                                    remaining_no_progress_retry_count -= 1;
+                                    if remaining_no_progress_retry_count == 0 {
+                                        return Err(SinkError::Kinesis(anyhow!(
+                                            "failed to send records. sent {} out of {}, last err: code: [{}], message: [{}]",
+                                            start_idx,
+                                            total_count,
+                                            result_entry.error_code.unwrap_or_default(),
+                                            result_entry.error_message.unwrap_or_default()
+                                        )));
+                                    } else {
+                                        warn!(
+                                            remaining_no_progress_retry_count,
+                                            sent = start_idx,
+                                            total_count,
+                                            "failed to send records. code: [{}], message: [{}]",
+                                            result_entry.error_code.unwrap_or_default(),
+                                            result_entry.error_message.unwrap_or_default()
+                                        )
+                                    }
+                                }
+                            } else {
+                                start_idx += record_count;
+                                // reset retry count when having progress
+                                remaining_no_progress_retry_count = MAX_NO_PROGRESS_RETRY_COUNT;
+                                // reset throttle delay when records can be fully sent.
+                                throttle_delay = None;
+                            }
+                        }
+                        Err(e) => {
+                            remaining_no_progress_retry_count -= 1;
+                            if remaining_no_progress_retry_count == 0 {
+                                #[expect(rw::format_error)]
+                                return Err(SinkError::Kinesis(anyhow!(
+                                    "failed to send records. sent {} out of {}, last err: {:?}",
+                                    start_idx,
+                                    total_count,
+                                    e.as_report()
+                                )));
```

Comment:
When wrapping an error and returning to caller (rather than handling/consuming the error as `warn!` below), use `context`:
```
return Err(
  SinkError::Kinesis(
    anyhow!(e).context(format!("failed to send records. sent {start_idx} out of {total_count}, last err:"))
  )
);
```